pkg/envoy/ads/stream.go

```diff
@@ -190,23 +193,62 @@ func respondToRequest(proxy *envoy.Proxy, discoveryRequest *xds_discovery.Discov
-		proxy.SetLastSentVersion(typeURL, ackVersion)
+	lastVersion = proxy.GetLastSentVersion(typeURL)
+	if requestVersion > lastVersion {
+		log.Debug().Msgf("Proxy SerialNumber=%s PodUID=%s: Higher version on request %s, req ver: %d - last ver: %d. Updating to match latest.",
+			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), typeURL.Short(), requestVersion, lastVersion)
+		proxy.SetLastSentVersion(typeURL, requestVersion)
+		return true
 	}
 
-	lastNonce := proxy.GetLastSentNonce(typeURL)
-	if lastNonce != "" && discoveryRequest.ResponseNonce == lastNonce {
-		log.Debug().Msgf("Proxy SerialNumber=%s PodUID=%s: Nothing changed for Envoy on Pod since Nonce=%s",
-			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), discoveryRequest.ResponseNonce)
+	// Compare Nonces
+	// As per protocol, we can ignore any request on the TypeURL stream that has not caught up with last sent nonce, if the
+	// nonce is non-empty.
+	// A newly connecting envoy might send a version and nonce from an old control plane, this is currently mitigated by the
+	// forced update @line 69
+	// TODO: Fix reconnection. Forced update causes other issues, we should stick to the protocol and wait for the first request
+	// and gracefully handle reconnecting envoys.
+	lastNonce = proxy.GetLastSentNonce(typeURL)
+	if requestNonce != lastNonce {
+		log.Debug().Msgf("Proxy SerialNumber=%s PodUID=%s: Ignoring request for %s non-latest nonce (request: %s, current: %s)",
+			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), typeURL.Short(), requestNonce, lastNonce)
 		return false
 	}
 
-	if discoveryRequest.ResponseNonce != "" {
-		log.Debug().Msgf("Proxy SerialNumber=%s PodUID=%s: Received discovery request with Nonce=%s matches=%t; proxy last Nonce=%s",
-			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), discoveryRequest.ResponseNonce, discoveryRequest.ResponseNonce == lastNonce, lastNonce)
+	// ----
+	// At this point, there is no error and nonces match, it is guaranteed an ACK with last version.
+	// What's left is to check if the resources listed are the same. If they are not, we must respond
+	// with the new reosurces requested
+	// This part of the code was inspired by Istio's `shouldRespond` handling of request resource difference
+	// https://github.com/istio/istio/blob/da6178604559bdf2c707a57f452d16bee0de90c8/pilot/pkg/xds/ads.go#L347
+	// ----
+	resourcesLastSent := proxy.GetLastResourcesSent(typeURL)
+	resourcesRequested := getRequestedResourceNamesSet(discoveryRequest)
+
+	// If what we last sent is a superset of what the
+	// requests resources subscribes to, it's ACK and nothing needs to be done.
+	// Otherwise, envoy might be asking us for additional resources that have to be sent along last time.
+	// Difference returns elemenets of <requested> that are not part of elements of <last sent>
+
+	requestedResourcesDifference := resourcesRequested.Difference(resourcesLastSent)
+	if requestedResourcesDifference.Cardinality() != 0 {
```

Comment:
Just need to ensure here Difference() works as expected:

Last sent: foo, bar, baz
Requested: foo

Request is ignored because `foo` was last sent.